src/libmongoc/tests/test-mongoc-write-concern.c

```diff
@@ -595,10 +596,137 @@ test_write_concern_wtimeout_preserved (void)
    ASSERT (bson_iter_next (&child));
    ASSERT_CMPSTR (bson_iter_key (&child), "wtimeout");
    ASSERT_CMPINT64 (bson_iter_int64 (&child), ==, INT64_MAX);
-   
+
    mongoc_write_concern_destroy (write_concern);
 }
 
+/* callback that records write concern for commands */
+static void
+write_concern_count (const mongoc_apm_command_started_t *event)
+{
+   const bson_t *command = mongoc_apm_command_started_get_command (event);
+   bson_iter_t iter, iter_rec;
+
+   if (bson_iter_init (&iter, command)) {
+      int *sent_collection_w =
+         (int *) mongoc_apm_command_started_get_context (event);
+      if (bson_iter_find_descendant (&iter, "writeConcern.w", &iter_rec)) {
+         *sent_collection_w = bson_iter_int32 (&iter_rec);
+      } else {
+         *sent_collection_w = MONGOC_WRITE_CONCERN_W_DEFAULT; /* no write
+                                                                 concern sent.
+                                                                 default used */
+      }
+   }
+}
+
+/* Addresses concerns brought up in CDRIVER-3595. This function comprises
+ * two tests. The first tests that with no txn in progress, wc is inherited
+ * from collection.  It also tests that an attempt to send an lsid with an
+ * unacknowledged wc fails.  The second tests that with a txn in progress, no
+ * wc is inherited from collection, unacknowledged or acknowledged. */
+void
+test_write_concern_inheritance_fam_txn (void *context)
+{
+   mongoc_find_and_modify_opts_t *opts = NULL;
+   bson_t *update;
+   bson_t *session_id;
+   bson_error_t error;
+   bson_t query = BSON_INITIALIZER;
+   mongoc_write_concern_t *wc;
+   mongoc_client_t *client;
+   mongoc_collection_t *collection;
+   mongoc_client_session_t *session = NULL;
+   mongoc_apm_callbacks_t *callbacks;
+   int sent_w = MONGOC_WRITE_CONCERN_W_DEFAULT;
+   bool success;
+   bool in_txn = *(bool *) context;
+
+   client = test_framework_client_new ();
+   collection = mongoc_client_get_collection (client, "db", "collection");
+
+   callbacks = mongoc_apm_callbacks_new ();
+   mongoc_apm_set_command_started_cb (callbacks, write_concern_count);
+   mongoc_client_set_apm_callbacks (client, callbacks, (void *) &sent_w);
+
+   wc = mongoc_write_concern_new ();
+   mongoc_write_concern_set_w (wc, 2);
+   mongoc_collection_set_write_concern (collection, wc);
+
+   BSON_APPEND_UTF8 (&query, "firstname", "Zlatan");
+   update = bson_new ();
+   session_id = bson_new ();
+
+   opts = mongoc_find_and_modify_opts_new ();
+   session = mongoc_client_start_session (client, NULL, &error);
+
```

Comment:
In the case where `in_txn` is false `session` is not immediately used, but it gets used later.  I did it this way because I wanted to test that unacknowledged and acknowledged write concerns are both inherited with no transaction in progress (and also no lsid in the command). It gets used later because in the `!in_txn` case I also test that sending an lsid with an unacknowledged write concern fails.

Reply:
Ah, in that case, I suggest further simplifying by parameterizing on both `in_txn` and `in_session`.  A struct of two bools could be used as the context.

```
typedef struct {
   bool in_session;
   bool in_txn;
} fam_inheritance_test_t;
```

The only minor annoyance is that the `test_write_concern_install` needs to create separate structs for each test.

```
fam_inheritance_test_t test_no_session = { false, false };
fam_inheritance_test_t test_with_session = { true, false };
fam_inheritance_test_t test_with_txn = { true, true };
/* ... */
```

But I think this further consolidates the test.